block-producer/src/block_builder/mod.rs

```diff
@@ -19,3 +45,99 @@ pub trait BlockBuilder: Send + Sync + 'static {
         batches: Vec<SharedTxBatch>,
     ) -> Result<(), BuildBlockError>;
 }
+
+#[derive(Debug)]
+pub struct DefaultBlockBuilder<S> {
+    store: Arc<S>,
+    account_root_program: AccountRootProgram,
+}
+
+impl<S> DefaultBlockBuilder<S>
+where
+    S: Store,
+{
+    pub fn new(store: Arc<S>) -> Self {
+        Self {
+            store,
+            account_root_program: AccountRootProgram::new(),
+        }
+    }
+}
+
+#[async_trait]
+impl<S> BlockBuilder for DefaultBlockBuilder<S>
+where
+    S: Store,
+{
+    async fn build_block(
+        &self,
+        batches: Vec<SharedTxBatch>,
+    ) -> Result<(), BuildBlockError> {
+        let updated_accounts: Vec<(AccountId, Digest)> =
+            batches.iter().flat_map(|batch| batch.updated_accounts()).collect();
+        let created_notes: Vec<Digest> =
+            batches.iter().flat_map(|batch| batch.created_notes()).collect();
+        let produced_nullifiers: Vec<Digest> =
+            batches.iter().flat_map(|batch| batch.produced_nullifiers()).collect();
+
+        let BlockInputs {
+            block_header: prev_block_header,
+            chain_peaks: _,
+            account_states: account_states_in_store,
+            nullifiers: _,
+        } = self
+            .store
+            .get_block_inputs(
+                updated_accounts.iter().map(|(account_id, _)| account_id),
+                produced_nullifiers.iter(),
+            )
+            .await
+            .unwrap();
```

Comment:
I changed it as follows:
+ `BlockHeaderWitness`: provides input to the VM program that generates the new header
+ `BlockKernel`: Encapsulates the VM program that computes the new header

The `BlockHeaderWitness` has a `into_parts(self) -> (AdviceInputs, StackInputs)` because it will only ever be called once by the `BlockKernel`. I think this is aligned with the direction you were trying to communicate in this review; what do you think?

The validation part is still TODO

Reply:
Yep - makes sense. I'd probably still call it just `BlockWitness` because it contains information needed to build a new block - but not a strong preference from my side.